pkgs/servers/mastodon/yarn-hash.nix

```diff
@@ -0,0 +1 @@
+"sha256-e3rl/WuKXaUdeDEYvo1sSubuIwtBjkbguCYdAijwXOA="
```

Comment:
When I rerun the update script, it actually writes here :+1:
It comes up with `1q2wy0l04796p3h4d3j11ciyxrjadj6vw61ig0fsapcadgyyaykv` though, without the `sha256` prefix.

Reply:
When I run the update script _within_ nixpkgs via `nix-shell ./maintainers/scripts/update.nix --argstr package mastodon` I also see this behavior 👍🏼 I'll update the `yarn-hash.nix` file accordingly.

What I can't figure out, and I'd been chalking up to my being very new to nix, is why [this usage](https://gitlab.com/tootcat/nixos-config/-/blob/79890506fb038a7727d6b9e4e2538314ab096411/update.sh#L29) seems to be running the update script from a different rev of nixpkgs (without the `yarn-hash.nix` generation) even though the [flake specifies this branch](https://gitlab.com/tootcat/nixos-config/-/blob/79890506fb038a7727d6b9e4e2538314ab096411/flake.nix#L5). What do you think?

Reply:
Are you sure you also use the latest revision? Your [flake.lock](https://gitlab.com/tootcat/nixos-config/-/blob/79890506fb038a7727d6b9e4e2538314ab096411/flake.lock) doesn't point to the current commit. You might need to run nix flake update --commit-lock-file first.

Reply:
OK, so I ran `nix flake update --commit-lock-file` which *did* update the [flake.lock](https://gitlab.com/tootcat/nixos-config/-/blob/3ee2d4510db0bc1d9a866e506584493d8914c454/flake.lock) but when running the update script I'm still seeing behavior *without* the generation of `yarn-hash.nix`, like the `mastodon.updateScript` in use is from a different rev 🤔

Reply:
Ah, now I see it. Your `update.sh` uses `nix-shell` in the shebang, which uses the channel mechanism to aquire Nixpkgs. That is the "old" way of doing things and runs completely in parallel to flakes.

Reply:
> When I run the update script _within_ nixpkgs via `nix-shell ./maintainers/scripts/update.nix --argstr package mastodon` I also see this behavior 👍🏼 I'll update the `yarn-hash.nix` file accordingly.

I take this back. Running `nix-shell ./maintainers/scripts/update.nix --argstr package mastodon` does not work for me as-is.

<details>
  <summary>script output</summary>

```
Going to be running update for following packages:
 - mastodon-4.1.7

Press Enter key to continue...

Running update for:
 - mastodon-4.1.7: UPDATING ...
 - mastodon-4.1.7: ERROR

--- SHOWING ERROR LOG FOR mastodon-4.1.7 ----------------------

error: hash 'null' does not include a type, nor is the type otherwise known from context
(use '--show-trace' to show detailed location information)


--- SHOWING ERROR LOG FOR mastodon-4.1.7 ----------------------
The update script for mastodon-4.1.7 failed with exit code 1
```

</details>

<details>
  <summary>mastodon.log</summary>

```
error: hash 'null' does not include a type, nor is the type otherwise known from context
(use '--show-trace' to show detailed location information)
```

</details>

<details>
  <summary>git diff after script fail</summary>

```diff
diff --git a/pkgs/servers/mastodon/source.nix b/pkgs/servers/mastodon/source.nix
index e0920eadf5f2..ab279088ab4d 100644
--- a/pkgs/servers/mastodon/source.nix
+++ b/pkgs/servers/mastodon/source.nix
@@ -4,7 +4,7 @@
     owner = "mastodon";
     repo = "mastodon";
     rev = "v4.1.7";
-    hash = "sha256-afGv7/Cl7bxbHvLXrtpZPlG4oJeL7LgQQKXBBt3LRSo=";
+    hash = "null";
   };
 in applyPatches {
   inherit src;
```

</details>

If I modify `pkgs/servers/mastodon/update.sh` like this, I get a successful update:

<details>
  <summary><code>pkgs/servers/mastodon/update.sh</code> diff</summary>

```diff
diff --git a/pkgs/servers/mastodon/update.sh b/pkgs/servers/mastodon/update.sh
index 1266770a4364..36669c2885cc 100755
--- a/pkgs/servers/mastodon/update.sh
+++ b/pkgs/servers/mastodon/update.sh
@@ -77,7 +77,7 @@ trap cleanup EXIT

 echo "Fetching source code $REVISION"
 JSON=$(nix-prefetch-github "$OWNER" "$REPO" --rev "$REVISION" 2> $WORK_DIR/nix-prefetch-git.out)
-HASH=$(echo "$JSON" | jq -r .hash)
+HASH=$(echo "$JSON" | jq -r .sha256)

 echo "Creating version.nix"
 echo "\"$VERSION\"" | sed 's/^"v/"/' > version.nix
@@ -89,7 +89,7 @@ cat > source.nix << EOF
     owner = "$OWNER";
     repo = "$REPO";
     rev = "$REVISION";
-    hash = "$HASH";
+    hash = "sha256-$HASH";
   };
 in applyPatches {
   inherit src;
```

</details>

Reply:
> Ah, now I see it. Your `update.sh` uses `nix-shell` in the shebang, which uses the channel mechanism to aquire Nixpkgs. That is the "old" way of doing things and runs completely in parallel to flakes.

Oh! TIL ❤️ Do you know offhand where I should be looking to learn how to change the `update.sh` to use flakes correctly? (fwiw I inherited this code 🤷🏼)

Reply:
Either put the dependencies into a flake `devShells` attribute and run the script within that shell, or you build it with Nix (e.g. with `pkgs.writeShellScriptBin "update" "your script content"`) as a flake output with all packages called explicitly (e.g. `${pkgs.yarn2nix}/bin/yarn2nix --bla`) and then run that via `nix run .#update`.